File: src/main/java/dominio/indicadores/RepositorioIndicadores.java
package dominio.indicadores;

import dominio.AbstractLocalRepository;
import dominio.empresas.Empresa;
import dominio.parser.ParserIndicadores;

import javax.persistence.Entity;
import java.time.Year;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Entity
public class RepositorioIndicadores extends AbstractLocalRepository<Indicador> {
	
	public void agregarMultiplesIndicadores(List<String> strIndicadores) {
		List<Indicador> indicadoresNuevos = obtenerIndicadoresParseados(strIndicadores);//NO ESTOY CATCHEANDO ParserError (!!!)
		indicadoresNuevos.forEach(ind -> agregar(ind));
	}
	
	public void eliminarIndicadores(){
		elementos.clear();
	}
	
	public void eliminarResultadosPrecalculados(){
		obtenerTodos().forEach(ind -> ind.eliminarResultadosPrecalculados());
	}

	@Override
	protected String mensajeEntidadExistenteError(Indicador elemento) {
		return "Ya existe un indicador con el nombre " + elemento.getNombre();
	}

	public List<Indicador> todosLosIndicadoresAplicablesA(Empresa empresa) {
		List<Indicador> indicadoresAplicables = new ArrayList<Indicador>();
		Set<Year> aniosDeCuentas = empresa.aniosDeLosQueTieneCuentas();
		aniosDeCuentas.forEach(anio -> indicadoresAplicables.addAll(this.indicadoresAplicablesA(empresa, anio)));
		return indicadoresAplicables;
	}

	public Set<Indicador> indicadoresAplicablesA(Empresa empresa, Year anio) {
		Set<Indicador> indicadoresAplicables = new HashSet<Indicador>();
		obtenerTodos().stream().filter(ind -> ind.esAplicableA(empresa, anio))
				.forEach(ind -> indicadoresAplicables.add(ind));
		return indicadoresAplicables;
	}

	public Indicador buscarIndicador(String nombreIndicador) {
		return obtenerTodos().stream().filter(ind -> ind.seLlama(nombreIndicador)).findFirst()
				.orElseThrow(() -> new NoExisteIndicadorError("No se pudo encontrar un indicador con ese nombre."));
	}
	
	private List<Indicador> obtenerIndicadoresParseados(List<String> strIndicadores){
		Stream<Indicador> indicadores = strIndicadores.stream().map(strInd -> ParserIndicadores.parse(strInd)).filter(ind -> !existe(ind));
		return indicadores.collect(Collectors.toList());
	}

}

class NoExisteIndicadorError extends RuntimeException {NoExisteIndicadorError(String e) {super(e);}}

File: src/main/java/dondeInvierto/dominio/repo/RepositorioIndicadores.kt
package dondeInvierto.dominio.repo

import dondeInvierto.dominio.empresas.Empresa
import dondeInvierto.dominio.indicadores.Indicador
import dondeInvierto.dominio.parser.ParserIndicadores
import java.time.Year
import javax.persistence.Entity

@Entity
class RepositorioIndicadores : AbstractLocalRepository<Indicador>() {

    fun agregarMultiplesIndicadores(strIndicadores: List<String>) {
        val indicadoresNuevos = obtenerIndicadoresParseados(strIndicadores)//NO ESTOY CATCHEANDO ParserError (!!!)
        indicadoresNuevos.forEach { ind -> agregar(ind) }
    }

    fun eliminarIndicadores() {
        elementos.clear()
    }

    fun eliminarResultadosPrecalculados() {
        obtenerTodos().forEach { ind -> ind.eliminarResultadosPrecalculados() }
    }

    override fun mensajeEntidadExistenteError(elemento: Indicador) = "Ya existe un indicador con el nombre " + elemento.nombre

    fun todosLosIndicadoresAplicablesA(empresa: Empresa): List<Indicador> {
        return empresa.aniosDeLosQueTieneCuentas()
                .flatMap { anio -> indicadoresAplicablesA(empresa, anio) }
    }

    fun indicadoresAplicablesA(empresa: Empresa, anio: Year): Set<Indicador> {
        return obtenerTodos().filter { it.esAplicableA(empresa, anio) }
                .toSet()
    }

    fun buscarIndicador(nombreIndicador: String): Indicador {
        return obtenerTodos().firstOrNull { it.seLlama(nombreIndicador) }
                ?: throw NoExisteIndicadorError("No se pudo encontrar un indicador con ese nombre.")
    }

    private fun obtenerIndicadoresParseados(strIndicadores: List<String>): List<Indicador> {
        return strIndicadores.map { ParserIndicadores.parse(it) }
                .filter { ind -> !existe(ind) }
    }

}

internal class NoExisteIndicadorError(e: String) : RuntimeException(e)